### ISSUE TITLE
Azkaban to have KILLING status transition

### ISSUE THREAD

**Issue by juhoautio** [azkaban/azkaban#1122]
We've had quite a lot of trouble with jobs/flows not terminating properly after execution has been killed. In that case it would be really helpful to see that the flow is still in "killing" state instead of "killed". Because if you kill an execution and expect it to have been terminated, but it hasn't, next day's scheduled run can be blocked by some job threads that were somehow hung (we mostly run everything in pipeline level 1 mode). Of course the real problem is somewhere else then (if a flow doesn't terminate properly). But this would at least give better visibility and save from most of the cases because if you notice that flow refuses to die you can restart the executor node.

Now it's just: RUNNING -> KILLED, although in reality some jobs can be still running, waiting for thread to exit, and what not.

So, add a new lifecycle phase so that it's: RUNNING -> KILLING -> KILLED.

I think it might be good if it's allowed to try killing the flow again while it's already being killed.. or if it makes more sense, while killing, azkaban could keep calling cancel() again on jobs that refuse to terminate, because it's possible that the first cancellation request failed for some reason.

From the perspective of blocking other jobs in pipeline mode, KILLING status should block newer executions in the same way that RUNNING status does.

**Comment by juhoautio** [comment]
@HappyRay & co., any chance to get some early feedback based on https://github.com/juhoautio/azkaban/pull/2? This is basically a working implementation, but it depends on some other PRs that I'd like to get in first. Then I will make a cleaned up & rebased PR based on https://github.com/juhoautio/azkaban/pull/2.

**Comment by HappyRay** [comment]
I am not against the idea of having this state given that administrators may choose to allow a relatively long timeout for a job to be terminated gracefully. 

As you suggested, we need to find a way to reliably terminate jobs and if we can't, notify the users. 

The current ProcessJob based jobs should force kill the job process see
azkaban.jobExecutor.ProcessJob#cancel

Do your jobs overwrite this method?

We typically avoid restarting executors since this will affect existing running flows. We start a new executor each time and leave the older executors running.

**Comment by HappyRay** [comment]
FYI since you use pipeline option, here is a related issue we discovered recently while reading the code. #1125 "The pipeline concurrent option is not reliable in the multiple executor mode"

**Comment by juhoautio** [comment]
Yes, we override cancel(). We don't also want to restart when it causes interrupting some "locally running jobs". But it may be sometimes the only option to clean things up if a thread is stuck (in a custom job type). Related to restarting, we're planning support for resuming flows and jobs automatically if executor node crashes.

**Comment by HappyRay** [comment]
I am interested in your thoughts on how to implement such a feature.
I hope to have the ability to reload executor processes without disrupting the existing flows.

**Comment by juhoautio** [comment]
Thanks for the info. Currently we run a single web node + a single executor node (per environment). But going to explore "rolling deployements" with multi executor setup. We don't really need multiple nodes to scale the jobs themselves because we execute heavy stuff on external systems (launched & tracked by azkaban).

**Comment by juhoautio** [comment]
We had a discussion about the resuming feature in gittter.im, azkaban/Lobby. Wanna join the chat?

**Comment by juhoautio** [comment]
I took the notes from gitter discussion and cleaned them up into this new issue: https://github.com/azkaban/azkaban/issues/1127

**Comment by juhoautio** [comment]
Resolved by https://github.com/azkaban/azkaban/pull/1509. Yay \o/
